catkin_tools/jobs/utils.py

```diff
@@ -50,42 +50,67 @@ def get_env_loaders(package, context):
     return sources
 
 
+def merge_envs(job_env, overlay_envs):
+    '''
+    In the merged/linked case of single env, this function amounts to a straight
+    assignment, but a more complex merge is required with isolated result spaces,
+    since a package's build environment may require extending that of multiple
+    other result spaces.
+    '''
+    merge_path_values = {}
+
+    for overlay_env in overlay_envs:
+        for key, values_str in overlay_env.items():
+            if key.endswith('PATH'):
```

Comment:
This function is used with and without `--env-cache`— it is required regardless because of the situation described in https://github.com/catkin/catkin_tools/issues/382#issuecomment-285739921. In short, each new `setup.bash` blows away the environment that was present before it, so this function is required any time there are multiple isolated result spaces which must be merged together to form the environment for a package's build (the only other way I can think of to deal with this would be to somehow mutate the various setup files to "chain up" the resultspaces, as if they had been built one after the other like with `catkin_make_isolated`).

So, we could turn this off for non-isolated-resultspace builds. We could also set an explicit whitelist of envvars for it to act on (`ROS_PACKAGE_PATH`, `CMAKE_PREFIX_PATH`, `CPATH`, `PYTHONPATH`, `PKG_CONFIG_PATH`, `PATH`). Or both. Let me know your preference.